File: src/main/java/com/saybetter/global/common/response/ResponseDto.java
package com.saybetter.global.common.response;

import org.springframework.http.HttpStatus;

import com.saybetter.global.common.code.BaseCode;
import com.saybetter.global.common.code.status.SuccessStatus;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;

@Getter
@AllArgsConstructor
@JsonPropertyOrder({"isSuccess", "code", "message", "result"})
public class ResponseDto<T> {

	@JsonProperty("isSuccess")
	private final Boolean isSuccess;
	private final String code;
	private final String message;

	@JsonInclude(JsonInclude.Include.NON_NULL)
	private final T result;

	// 성공한 경우
	public static <T> ResponseDto<T> onSuccess(T result) {
		return new ResponseDto<>(
				true,
				SuccessStatus.OK.getCode(),
				SuccessStatus.OK.getMessage(),
				result
		);
	}

	public static <T> ResponseDto<T> of(BaseCode code, T result) {
		return new ResponseDto<>(
				true,
				code.getReasonHttpStatus().getCode(),
				code.getReasonHttpStatus().getMessage(),
				result
		);
	}

	// 실패한 경우
	public static <T> ResponseDto<T> onFailure(String code, String message, T result) {
		return new ResponseDto<>(
				false,
				code,
				message,
				result
		);
	}

	@Getter
	@Builder
	public static class ErrorReasonDto {
		private HttpStatus httpStatus;
		private final boolean isSuccess;
		private final String code;
		private final String message;
	}

	@Getter
	@Builder
	public static class ReasonDto {
		private HttpStatus httpStatus;
		private final boolean isSuccess;
		private final String code;
		private final String message;
	}
}

File: core/core-api/src/main/kotlin/io/say/better/global/common/response/ResponseDto.kt
package io.say.better.global.common.response

import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonPropertyOrder
import io.say.better.global.common.code.BaseCode
import io.say.better.global.common.code.status.SuccessStatus
import lombok.AllArgsConstructor
import lombok.Builder
import lombok.Getter
import org.springframework.http.HttpStatus

@Getter
@AllArgsConstructor
@JsonPropertyOrder("isSuccess", "code", "message", "result")
data class ResponseDto<T>(
        @field:JsonProperty("isSuccess")
        private val isSuccess: Boolean,
        @field:JsonProperty("code")
        private val code: String,
        @field:JsonProperty("message")
        private val message: String,

        @field:JsonInclude(JsonInclude.Include.NON_NULL)
        @field:JsonProperty("result")
        private val result: T,
) {

    data class ErrorReasonDto(
            val httpStatus: HttpStatus? = null,
            val isSuccess: Boolean = false,
            val code: String,
            val message: String
    )

    data class ReasonDto(
            val httpStatus: HttpStatus? = null,
            val isSuccess: Boolean = false,
            val code: String,
            val message: String
    )

    companion object {
        // 성공한 경우
        fun <T> onSuccess(result: T): ResponseDto<T> {
            return ResponseDto(
                    isSuccess = true,
                    code = SuccessStatus.OK.code,
                    message = SuccessStatus.OK.message,
                    result = result
            )
        }

        fun <T> of(code: BaseCode, result: T): ResponseDto<T> {
            return ResponseDto(
                    isSuccess = true,
                    code = code.reasonHttpStatus.code,
                    message = code.reasonHttpStatus.message,
                    result = result
            )
        }

        // 실패한 경우
        fun <T> onFailure(code: String, message: String, result: T): ResponseDto<T> {
            return ResponseDto(
                    isSuccess = false,
                    code = code,
                    message = message,
                    result = result
            )
        }
    }
}
